internal/wasm/sdk/internal/wasm/bindings.go

```diff
@@ -4,93 +4,196 @@
+	exports := getExports(caller)
+
+	var convertedArgs []*ast.Term
+
+	// first two args are the built-in identifier and context structure
+	for i := 2; i < len(args); i++ {
+
+		x, err := fromWasmValue(exports, args[i].I32())
+		if err != nil {
+			panic(builtinError{err: err})
+		}
+
+		convertedArgs = append(convertedArgs, x)
+	}
+
+	var output *ast.Term
+
+	if err := d.builtins[args[0].I32()](*d.ctx, convertedArgs, func(t *ast.Term) error {
+		output = t
+		return nil
+	}); err != nil {
+		panic(builtinError{err: err})
+	}
+
+	// if output is undefined, return NULL
+	if output == nil {
+		return []wasmtime.Val{wasmtime.ValI32(0)}, nil
+	}
+
+	addr, err := toWasmValue(exports, output)
 	if err != nil {
-		return nil, err
+		panic(builtinError{err: err})
 	}
 
-	return imports.AppendFunction("opa_builtin4", opa_builtin4, C.opa_builtin4)
+	return []wasmtime.Val{wasmtime.ValI32(addr)}, nil
 }
 
-//export opa_println
-func opa_println(ctx unsafe.Pointer, addr int32) {
-	getVM(ctx).Println(addr)
+type exports struct {
+	Memory       *wasmtime.Memory
+	mallocFn     *wasmtime.Func
+	valueDumpFn  *wasmtime.Func
+	valueParseFn *wasmtime.Func
 }
 
-//export opa_abort
-func opa_abort(ctx unsafe.Pointer, addr int32) {
-	getVM(ctx).Abort(addr)
+func getExports(c *wasmtime.Caller) exports {
+	var e exports
+	e.Memory = c.GetExport("memory").Memory()
+	e.mallocFn = c.GetExport("opa_malloc").Func()
+	e.valueDumpFn = c.GetExport("opa_value_dump").Func()
+	e.valueParseFn = c.GetExport("opa_value_parse").Func()
+	return e
 }
 
-//export opa_builtin0
-func opa_builtin0(ctx unsafe.Pointer, builtinID, context int32) int32 {
-	return getVM(ctx).Builtin(builtinID, context)
+func (e exports) Malloc(len int32) (int32, error) {
+	ptr, err := e.mallocFn.Call(len)
+	if err != nil {
+		return 0, err
+	}
+	return ptr.(int32), nil
 }
 
-//export opa_builtin1
-func opa_builtin1(ctx unsafe.Pointer, builtinID, context, arg0 int32) int32 {
-	return getVM(ctx).Builtin(builtinID, context, arg0)
+func (e exports) ValueDump(addr int32) (int32, error) {
+	result, err := e.valueDumpFn.Call(addr)
+	if err != nil {
+		return 0, err
+	}
+	return result.(int32), nil
 }
 
-//export opa_builtin2
-func opa_builtin2(ctx unsafe.Pointer, builtinID, context, arg0, arg1 int32) int32 {
-	return getVM(ctx).Builtin(builtinID, context, arg0, arg1)
+func (e exports) ValueParse(addr int32, len int32) (int32, error) {
+	result, err := e.valueParseFn.Call(addr, len)
+	if err != nil {
+		return 0, err
+	}
+	return result.(int32), nil
 }
 
-//export opa_builtin3
-func opa_builtin3(ctx unsafe.Pointer, builtinID, context, arg0, arg1, arg2 int32) int32 {
-	return getVM(ctx).Builtin(builtinID, context, arg0, arg1, arg2)
-}
+func fromWasmValue(e exports, addr int32) (*ast.Term, error) {
+
+	serialized, err := e.ValueDump(addr)
+	if err != nil {
+		return nil, err
+	}
+
+	data := e.Memory.UnsafeData()[serialized:]
+	n := bytes.IndexByte(data, 0)
+	if n < 0 {
+		n = 0
```

Comment:
I believe we've talked about this being less than ideal and returning an error might be better (since `ast.ParseTerm` wouldn't like `""`)... now is as good a time as any to do so, I guess.

Reply:
Agreed. Replaced with an error.